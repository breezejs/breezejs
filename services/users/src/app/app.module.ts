/**
 * @author Mo Gusbi <me@mogusbi.co.uk>
 */
import {Module} from '@nestjs/common';
import {UsersModule} from './users';

/**
 * App module for users service
 */
@Module({
  imports: [
    UsersModule
  ]
})
export class AppModule {}
