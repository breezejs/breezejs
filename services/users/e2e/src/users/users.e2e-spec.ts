/**
 * @author Mo Gusbi <me@mogusbi.co.uk>
 */
import {INestApplication, ValidationPipe} from '@nestjs/common';
import {Test, TestingModule} from '@nestjs/testing';
import * as supertest from 'supertest';
import {UsersModule} from '../../../src/app/users';

describe('Users', (): void => {
  let app: INestApplication;
  let id: string;
  let request: supertest.SuperTest<supertest.Request>;

  beforeAll(async (): Promise<void> => {
    const testModule: TestingModule = await Test
      .createTestingModule({
        imports: [
          UsersModule
        ]
      })
      .compile();

    app = testModule.createNestApplication();

    app.useGlobalPipes(new ValidationPipe());

    await app.init();

    request = supertest(app.getHttpServer());
  });

  afterAll(async (): Promise<void> => {
    await app.close();
  });

  describe('[GET] /users', (): void => {
    describe('with no filtering or sorting', (): void => {
      it('should list first page of users with all values', async (): Promise<void> => {
        const {body, status}: supertest.Response = await request.get('/users');

        expect(status).toEqual(200);
        expect(body).toMatchSnapshot();
      });

      it('should list second page of users with all values', async (): Promise<void> => {
        const {body, status}: supertest.Response = await request.get('/users').query({
          page: 2
        });

        expect(status).toEqual(200);
        expect(body).toMatchSnapshot();
      });

      it('should return no content if out of bounds', async (): Promise<void> => {
        const {body, status}: supertest.Response = await request.get('/users').query({
          page: 3
        });

        expect(status).toEqual(204);
        expect(body).toMatchSnapshot();
      });
    });

    describe('with filtering', (): void => {
      it('should list first page of users with selected values', async (): Promise<void> => {
        const {body, status}: supertest.Response = await request.get('/users').query({
          fields: 'user.id,user.forename,user.surname'
        });

        expect(status).toEqual(200);
        expect(body).toMatchSnapshot();
      });

      it('should list second page of users with selected values', async (): Promise<void> => {
        const {body, status}: supertest.Response = await request.get('/users').query({
          fields: 'user.id,user.forename,user.surname',
          page: 2
        });

        expect(status).toEqual(200);
        expect(body).toMatchSnapshot();
      });

      it('should return no content if out of bounds', async (): Promise<void> => {
        const {body, status}: supertest.Response = await request.get('/users').query({
          fields: 'user.id,user.forename,user.surname',
          page: 3
        });

        expect(status).toEqual(204);
        expect(body).toMatchSnapshot();
      });

      it('should return a bad request if filtered field does not exist', async (): Promise<void> => {
        const {body, status}: supertest.Response = await request.get('/users').query({
          fields: 'user.id,user.name,user.surname'
        });

        expect(status).toEqual(400);
        expect(body).toMatchSnapshot();
      });
    });

    describe('with sorting', (): void => {
      describe('ascending', (): void => {
        it('should list first page of users with all values', async (): Promise<void> => {
          const {body, status}: supertest.Response = await request.get('/users').query({
            dir: 'asc',
            sort: 'surname'
          });

          expect(status).toEqual(200);
          expect(body).toMatchSnapshot();
        });

        it('should list second page of users with all values', async (): Promise<void> => {
          const {body, status}: supertest.Response = await request.get('/users').query({
            dir: 'asc',
            page: 2,
            sort: 'surname'
          });

          expect(status).toEqual(200);
          expect(body).toMatchSnapshot();
        });

        it('should return no content if out of bounds', async (): Promise<void> => {
          const {body, status}: supertest.Response = await request.get('/users').query({
            dir: 'asc',
            page: 3,
            sort: 'surname'
          });

          expect(status).toEqual(204);
          expect(body).toMatchSnapshot();
        });
      });

      describe('descending', (): void => {
        it('should list first page of users with all values', async (): Promise<void> => {
          const {body, status}: supertest.Response = await request.get('/users').query({
            dir: 'desc',
            sort: 'surname'
          });

          expect(status).toEqual(200);
          expect(body).toMatchSnapshot();
        });

        it('should list second page of users with all values', async (): Promise<void> => {
          const {body, status}: supertest.Response = await request.get('/users').query({
            dir: 'desc',
            page: 2,
            sort: 'surname'
          });

          expect(status).toEqual(200);
          expect(body).toMatchSnapshot();
        });

        it('should return no content if out of bounds', async (): Promise<void> => {
          const {body, status}: supertest.Response = await request.get('/users').query({
            dir: 'desc',
            page: 3,
            sort: 'surname'
          });

          expect(status).toEqual(204);
          expect(body).toMatchSnapshot();
        });
      });
    });

    describe('with limiting', (): void => {
      it('should list first page of users with all values', async (): Promise<void> => {
        const {body, status}: supertest.Response = await request.get('/users').query({
          limit: 5
        });

        expect(status).toEqual(200);
        expect(body).toMatchSnapshot();
      });

      it('should list second page of users with all values', async (): Promise<void> => {
        const {body, status}: supertest.Response = await request.get('/users').query({
          limit: 5,
          page: 2
        });

        expect(status).toEqual(200);
        expect(body).toMatchSnapshot();
      });

      it('should return no content if out of bounds', async (): Promise<void> => {
        const {body, status}: supertest.Response = await request.get('/users').query({
          limit: 5,
          page: 4
        });

        expect(status).toEqual(204);
        expect(body).toMatchSnapshot();
      });
    });
  });

  describe('[GET] /users/:id', (): void => {
    it('should return user', async (): Promise<void> => {
      const {body, status}: supertest.Response = await request.get('/users/f07a1ccb-f3bf-4ac8-beab-c740668674c4');

      expect(status).toEqual(200);
      expect(body).toMatchSnapshot();
    });

    it('should return a filtered result', async (): Promise<void> => {
      const {body, status}: supertest.Response = await request.get('/users/f07a1ccb-f3bf-4ac8-beab-c740668674c4').query({
        fields: 'user.id,user.emailAddress'
      });

      expect(status).toEqual(200);
      expect(body).toMatchSnapshot();
    });

    it('should return a bad request if filtered field does not exist', async (): Promise<void> => {
      const {body, status}: supertest.Response = await request.get('/users/f07a1ccb-f3bf-4ac8-beab-c740668674c4').query({
        fields: 'user.id,user.name'
      });

      expect(status).toEqual(400);
      expect(body).toMatchSnapshot();
    });

    it('should return not found if user does not exist', async (): Promise<void> => {
      const {body, status}: supertest.Response = await request.get('/users/f07a1ccb-f3bf-4ac8-beab-c740668674c9');

      expect(status).toEqual(404);
      expect(body).toMatchSnapshot();
    });
  });

  describe('[POST] /users', (): void => {
    it('should create a new user', async (): Promise<void> => {
      const {body, status}: supertest.Response = await request.post('/users').send({
        emailAddress: 'test@example.com',
        forename: 'Test',
        surname: 'Example'
      });

      id = body.id;

      expect(status).toEqual(201);
      expect(body).toEqual({
        createdAt: expect.any(String),
        emailAddress: 'test@example.com',
        forename: 'Test',
        id: expect.any(String),
        surname: 'Example',
        updatedAt: expect.any(String)
      });
    });

    it('should throw an error if a duplicate user is created', async (): Promise<void> => {
      const {body, status}: supertest.Response = await request.post('/users').send({
        emailAddress: 'test@example.com',
        forename: 'Duplicate',
        surname: 'User'
      });

      expect(status).toEqual(400);
      expect(body).toMatchSnapshot();
    });

    it('should throw an error if a required field is missing', async (): Promise<void> => {
      const {body, status}: supertest.Response = await request.post('/users').send({
        forename: 'Duplicate',
        surname: 'User'
      });

      expect(status).toEqual(400);
      expect(body).toMatchSnapshot();
    });

    it('should throw an error if a field is the wrong format', async (): Promise<void> => {
      const {body, status}: supertest.Response = await request.post('/users').send({
        emailAddress: 'test[AT]example.com',
        forename: 'Duplicate',
        surname: 'User'
      });

      expect(status).toEqual(400);
      expect(body).toMatchSnapshot();
    });
  });

  describe('[PATCH] /users/:id', (): void => {
    it('should update a user', async (): Promise<void> => {
      const {body, status}: supertest.Response = await request.patch(`/users/${id}`).send({
        emailAddress: 'updated@example.com',
        forename: 'Updated',
        surname: 'Example'
      });

      expect(status).toEqual(204);
      expect(body).toMatchSnapshot();
    });

    it('should throw an error if a duplicate user is created', async (): Promise<void> => {
      const {body, status}: supertest.Response = await request.patch(`/users/${id}`).send({
        emailAddress: 'me@mogusbi.co.uk',
        forename: 'Test',
        surname: 'User'
      });

      expect(status).toEqual(400);
      expect(body).toMatchSnapshot();
    });

    it('should throw an error if a required field is missing', async (): Promise<void> => {
      const {body, status}: supertest.Response = await request.patch(`/users/${id}`).send({
        forename: 'Duplicate',
        surname: 'User'
      });

      expect(status).toEqual(400);
      expect(body).toMatchSnapshot();
    });

    it('should throw an error if a field is the wrong format', async (): Promise<void> => {
      const {body, status}: supertest.Response = await request.patch(`/users/${id}`).send({
        emailAddress: 'test[AT]example.com',
        forename: 'Duplicate',
        surname: 'User'
      });

      expect(status).toEqual(400);
      expect(body).toMatchSnapshot();
    });
  });

  describe('[DELETE] /users', (): void => {
    it('should delete user', async (): Promise<void> => {
      const {body, status}: supertest.Response = await request.delete(`/users/${id}`);

      expect(status).toEqual(204);
      expect(body).toMatchSnapshot();
    });

    it('should return not found if user does not exist', async (): Promise<void> => {
      const {body, status}: supertest.Response = await request.delete('/users/88044266-a83a-4e10-b85b-e552790909a1');

      expect(status).toEqual(404);
      expect(body).toMatchSnapshot();
    });
  });
});
