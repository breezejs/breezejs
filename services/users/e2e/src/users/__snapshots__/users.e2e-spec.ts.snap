// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users [DELETE] /users should delete user 1`] = `Object {}`;

exports[`Users [DELETE] /users should return not found if user does not exist 1`] = `
Object {
  "error": "Not Found",
  "statusCode": 404,
}
`;

exports[`Users [GET] /users with filtering should list first page of users with selected values 1`] = `
Object {
  "items": Array [
    Object {
      "forename": "Jeremy",
      "id": "034682ae-81e6-4c13-b376-503d5b11fd33",
      "surname": "Rice",
    },
    Object {
      "forename": "Andrea",
      "id": "1ea35773-2d0e-42ab-bd2a-c59fa47cca7e",
      "surname": "Horton",
    },
    Object {
      "forename": "Erin",
      "id": "2b56ba98-b120-4d7a-a56f-0225c07218f4",
      "surname": "Wade",
    },
    Object {
      "forename": "Dale",
      "id": "33ccca6b-1a2d-412e-811a-0937648a18d2",
      "surname": "Love",
    },
    Object {
      "forename": "Miranda",
      "id": "4a517c57-731e-4fd4-b67a-407c7bcfd2bf",
      "surname": "Thompson",
    },
    Object {
      "forename": "Jeremiah",
      "id": "6250affa-690b-498d-94ff-8067600a85d4",
      "surname": "Hubbard",
    },
    Object {
      "forename": "Emanuel",
      "id": "8793290b-95aa-45a1-b022-e997d3144422",
      "surname": "Wong",
    },
    Object {
      "forename": "Oliver",
      "id": "923ac85f-5bf1-4902-8946-eaebe76261cb",
      "surname": "Massey",
    },
    Object {
      "forename": "Eric",
      "id": "98dc24a5-4f9a-4013-8124-3843ad1278d2",
      "surname": "Allen",
    },
    Object {
      "forename": "Inez",
      "id": "a5eaa098-d89f-4dde-8cc7-1c410086d922",
      "surname": "Campbell",
    },
  ],
  "limit": 10,
  "page": 1,
  "pages": 2,
  "total": 12,
}
`;

exports[`Users [GET] /users with filtering should list second page of users with selected values 1`] = `
Object {
  "items": Array [
    Object {
      "forename": "Mo",
      "id": "f07a1ccb-f3bf-4ac8-beab-c740668674c4",
      "surname": "Gusbi",
    },
    Object {
      "forename": "Jessie",
      "id": "fba98a9c-65c4-40d4-b9ad-0187886dabbd",
      "surname": "Harrington",
    },
  ],
  "limit": 10,
  "page": 2,
  "pages": 2,
  "total": 12,
}
`;

exports[`Users [GET] /users with filtering should return a bad request if filtered field does not exist 1`] = `
Object {
  "error": "Bad Request",
  "message": "ER_BAD_FIELD_ERROR: Unknown column 'user.name' in 'field list'",
  "statusCode": 400,
}
`;

exports[`Users [GET] /users with filtering should return no content if out of bounds 1`] = `Object {}`;

exports[`Users [GET] /users with limiting should list first page of users with all values 1`] = `
Object {
  "items": Array [
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Jeremy.Rice@mailinator.com",
      "forename": "Jeremy",
      "id": "034682ae-81e6-4c13-b376-503d5b11fd33",
      "surname": "Rice",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Andrea.Horton@mailinator.com",
      "forename": "Andrea",
      "id": "1ea35773-2d0e-42ab-bd2a-c59fa47cca7e",
      "surname": "Horton",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Erin.Wade@mailinator.com",
      "forename": "Erin",
      "id": "2b56ba98-b120-4d7a-a56f-0225c07218f4",
      "surname": "Wade",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Dale.Love@mailinator.com",
      "forename": "Dale",
      "id": "33ccca6b-1a2d-412e-811a-0937648a18d2",
      "surname": "Love",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Miranda.Thompson@mailinator.com",
      "forename": "Miranda",
      "id": "4a517c57-731e-4fd4-b67a-407c7bcfd2bf",
      "surname": "Thompson",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
  ],
  "limit": 5,
  "page": 1,
  "pages": 3,
  "total": 12,
}
`;

exports[`Users [GET] /users with limiting should list second page of users with all values 1`] = `
Object {
  "items": Array [
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Jeremiah.Hubbard@mailinator.com",
      "forename": "Jeremiah",
      "id": "6250affa-690b-498d-94ff-8067600a85d4",
      "surname": "Hubbard",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Emanuel.Wong@mailinator.com",
      "forename": "Emanuel",
      "id": "8793290b-95aa-45a1-b022-e997d3144422",
      "surname": "Wong",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Oliver.Massey@mailinator.com",
      "forename": "Oliver",
      "id": "923ac85f-5bf1-4902-8946-eaebe76261cb",
      "surname": "Massey",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Eric.Allen@mailinator.com",
      "forename": "Eric",
      "id": "98dc24a5-4f9a-4013-8124-3843ad1278d2",
      "surname": "Allen",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Inez.Campbell@mailinator.com",
      "forename": "Inez",
      "id": "a5eaa098-d89f-4dde-8cc7-1c410086d922",
      "surname": "Campbell",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
  ],
  "limit": 5,
  "page": 2,
  "pages": 3,
  "total": 12,
}
`;

exports[`Users [GET] /users with limiting should return no content if out of bounds 1`] = `Object {}`;

exports[`Users [GET] /users with no filtering or sorting should list first page of users with all values 1`] = `
Object {
  "items": Array [
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Jeremy.Rice@mailinator.com",
      "forename": "Jeremy",
      "id": "034682ae-81e6-4c13-b376-503d5b11fd33",
      "surname": "Rice",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Andrea.Horton@mailinator.com",
      "forename": "Andrea",
      "id": "1ea35773-2d0e-42ab-bd2a-c59fa47cca7e",
      "surname": "Horton",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Erin.Wade@mailinator.com",
      "forename": "Erin",
      "id": "2b56ba98-b120-4d7a-a56f-0225c07218f4",
      "surname": "Wade",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Dale.Love@mailinator.com",
      "forename": "Dale",
      "id": "33ccca6b-1a2d-412e-811a-0937648a18d2",
      "surname": "Love",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Miranda.Thompson@mailinator.com",
      "forename": "Miranda",
      "id": "4a517c57-731e-4fd4-b67a-407c7bcfd2bf",
      "surname": "Thompson",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Jeremiah.Hubbard@mailinator.com",
      "forename": "Jeremiah",
      "id": "6250affa-690b-498d-94ff-8067600a85d4",
      "surname": "Hubbard",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Emanuel.Wong@mailinator.com",
      "forename": "Emanuel",
      "id": "8793290b-95aa-45a1-b022-e997d3144422",
      "surname": "Wong",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Oliver.Massey@mailinator.com",
      "forename": "Oliver",
      "id": "923ac85f-5bf1-4902-8946-eaebe76261cb",
      "surname": "Massey",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Eric.Allen@mailinator.com",
      "forename": "Eric",
      "id": "98dc24a5-4f9a-4013-8124-3843ad1278d2",
      "surname": "Allen",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Inez.Campbell@mailinator.com",
      "forename": "Inez",
      "id": "a5eaa098-d89f-4dde-8cc7-1c410086d922",
      "surname": "Campbell",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
  ],
  "limit": 10,
  "page": 1,
  "pages": 2,
  "total": 12,
}
`;

exports[`Users [GET] /users with no filtering or sorting should list second page of users with all values 1`] = `
Object {
  "items": Array [
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "me@mogusbi.co.uk",
      "forename": "Mo",
      "id": "f07a1ccb-f3bf-4ac8-beab-c740668674c4",
      "surname": "Gusbi",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Jessie.Harrington@mailinator.com",
      "forename": "Jessie",
      "id": "fba98a9c-65c4-40d4-b9ad-0187886dabbd",
      "surname": "Harrington",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
  ],
  "limit": 10,
  "page": 2,
  "pages": 2,
  "total": 12,
}
`;

exports[`Users [GET] /users with no filtering or sorting should return no content if out of bounds 1`] = `Object {}`;

exports[`Users [GET] /users with sorting ascending should list first page of users with all values 1`] = `
Object {
  "items": Array [
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Eric.Allen@mailinator.com",
      "forename": "Eric",
      "id": "98dc24a5-4f9a-4013-8124-3843ad1278d2",
      "surname": "Allen",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Inez.Campbell@mailinator.com",
      "forename": "Inez",
      "id": "a5eaa098-d89f-4dde-8cc7-1c410086d922",
      "surname": "Campbell",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "me@mogusbi.co.uk",
      "forename": "Mo",
      "id": "f07a1ccb-f3bf-4ac8-beab-c740668674c4",
      "surname": "Gusbi",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Jessie.Harrington@mailinator.com",
      "forename": "Jessie",
      "id": "fba98a9c-65c4-40d4-b9ad-0187886dabbd",
      "surname": "Harrington",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Andrea.Horton@mailinator.com",
      "forename": "Andrea",
      "id": "1ea35773-2d0e-42ab-bd2a-c59fa47cca7e",
      "surname": "Horton",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Jeremiah.Hubbard@mailinator.com",
      "forename": "Jeremiah",
      "id": "6250affa-690b-498d-94ff-8067600a85d4",
      "surname": "Hubbard",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Dale.Love@mailinator.com",
      "forename": "Dale",
      "id": "33ccca6b-1a2d-412e-811a-0937648a18d2",
      "surname": "Love",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Oliver.Massey@mailinator.com",
      "forename": "Oliver",
      "id": "923ac85f-5bf1-4902-8946-eaebe76261cb",
      "surname": "Massey",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Jeremy.Rice@mailinator.com",
      "forename": "Jeremy",
      "id": "034682ae-81e6-4c13-b376-503d5b11fd33",
      "surname": "Rice",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Miranda.Thompson@mailinator.com",
      "forename": "Miranda",
      "id": "4a517c57-731e-4fd4-b67a-407c7bcfd2bf",
      "surname": "Thompson",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
  ],
  "limit": 10,
  "page": 1,
  "pages": 2,
  "total": 12,
}
`;

exports[`Users [GET] /users with sorting ascending should list second page of users with all values 1`] = `
Object {
  "items": Array [
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Erin.Wade@mailinator.com",
      "forename": "Erin",
      "id": "2b56ba98-b120-4d7a-a56f-0225c07218f4",
      "surname": "Wade",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Emanuel.Wong@mailinator.com",
      "forename": "Emanuel",
      "id": "8793290b-95aa-45a1-b022-e997d3144422",
      "surname": "Wong",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
  ],
  "limit": 10,
  "page": 2,
  "pages": 2,
  "total": 12,
}
`;

exports[`Users [GET] /users with sorting ascending should return no content if out of bounds 1`] = `Object {}`;

exports[`Users [GET] /users with sorting descending should list first page of users with all values 1`] = `
Object {
  "items": Array [
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Emanuel.Wong@mailinator.com",
      "forename": "Emanuel",
      "id": "8793290b-95aa-45a1-b022-e997d3144422",
      "surname": "Wong",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Erin.Wade@mailinator.com",
      "forename": "Erin",
      "id": "2b56ba98-b120-4d7a-a56f-0225c07218f4",
      "surname": "Wade",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Miranda.Thompson@mailinator.com",
      "forename": "Miranda",
      "id": "4a517c57-731e-4fd4-b67a-407c7bcfd2bf",
      "surname": "Thompson",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Jeremy.Rice@mailinator.com",
      "forename": "Jeremy",
      "id": "034682ae-81e6-4c13-b376-503d5b11fd33",
      "surname": "Rice",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Oliver.Massey@mailinator.com",
      "forename": "Oliver",
      "id": "923ac85f-5bf1-4902-8946-eaebe76261cb",
      "surname": "Massey",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Dale.Love@mailinator.com",
      "forename": "Dale",
      "id": "33ccca6b-1a2d-412e-811a-0937648a18d2",
      "surname": "Love",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Jeremiah.Hubbard@mailinator.com",
      "forename": "Jeremiah",
      "id": "6250affa-690b-498d-94ff-8067600a85d4",
      "surname": "Hubbard",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Andrea.Horton@mailinator.com",
      "forename": "Andrea",
      "id": "1ea35773-2d0e-42ab-bd2a-c59fa47cca7e",
      "surname": "Horton",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Jessie.Harrington@mailinator.com",
      "forename": "Jessie",
      "id": "fba98a9c-65c4-40d4-b9ad-0187886dabbd",
      "surname": "Harrington",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "me@mogusbi.co.uk",
      "forename": "Mo",
      "id": "f07a1ccb-f3bf-4ac8-beab-c740668674c4",
      "surname": "Gusbi",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
  ],
  "limit": 10,
  "page": 1,
  "pages": 2,
  "total": 12,
}
`;

exports[`Users [GET] /users with sorting descending should list second page of users with all values 1`] = `
Object {
  "items": Array [
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Inez.Campbell@mailinator.com",
      "forename": "Inez",
      "id": "a5eaa098-d89f-4dde-8cc7-1c410086d922",
      "surname": "Campbell",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
    Object {
      "createdAt": "2019-02-10T00:00:00.000Z",
      "emailAddress": "Eric.Allen@mailinator.com",
      "forename": "Eric",
      "id": "98dc24a5-4f9a-4013-8124-3843ad1278d2",
      "surname": "Allen",
      "updatedAt": "2019-02-10T00:00:00.000Z",
    },
  ],
  "limit": 10,
  "page": 2,
  "pages": 2,
  "total": 12,
}
`;

exports[`Users [GET] /users with sorting descending should return no content if out of bounds 1`] = `Object {}`;

exports[`Users [GET] /users/:id should return a bad request if filtered field does not exist 1`] = `
Object {
  "error": "Bad Request",
  "message": "ER_BAD_FIELD_ERROR: Unknown column 'user.name' in 'field list'",
  "statusCode": 400,
}
`;

exports[`Users [GET] /users/:id should return a filtered result 1`] = `
Object {
  "emailAddress": "me@mogusbi.co.uk",
  "id": "f07a1ccb-f3bf-4ac8-beab-c740668674c4",
}
`;

exports[`Users [GET] /users/:id should return not found if fixture does not exist 1`] = `
Object {
  "error": "Not Found",
  "statusCode": 404,
}
`;

exports[`Users [GET] /users/:id should return not found if user does not exist 1`] = `
Object {
  "error": "Not Found",
  "statusCode": 404,
}
`;

exports[`Users [GET] /users/:id should return user 1`] = `
Object {
  "createdAt": "2019-02-10T00:00:00.000Z",
  "emailAddress": "me@mogusbi.co.uk",
  "forename": "Mo",
  "id": "f07a1ccb-f3bf-4ac8-beab-c740668674c4",
  "surname": "Gusbi",
  "updatedAt": "2019-02-10T00:00:00.000Z",
}
`;

exports[`Users [PATCH] /users/:id should throw an error if a duplicate user is created 1`] = `
Object {
  "error": "Bad Request",
  "message": "ER_DUP_ENTRY: Duplicate entry 'me@mogusbi.co.uk' for key 'IDX_eea9ba2f6e1bb8cb89c4e672f6'",
  "statusCode": 400,
}
`;

exports[`Users [PATCH] /users/:id should throw an error if a field is the wrong format 1`] = `
Array [
  Object {
    "children": Array [],
    "constraints": Object {
      "isEmail": "emailAddress must be an email",
    },
    "property": "emailAddress",
    "target": Object {
      "emailAddress": "test[AT]example.com",
      "forename": "Duplicate",
      "surname": "User",
    },
    "value": "test[AT]example.com",
  },
]
`;

exports[`Users [PATCH] /users/:id should throw an error if a required field is missing 1`] = `
Array [
  Object {
    "children": Array [],
    "constraints": Object {
      "isEmail": "emailAddress must be an email",
      "isNotEmpty": "emailAddress should not be empty",
    },
    "property": "emailAddress",
    "target": Object {
      "forename": "Duplicate",
      "surname": "User",
    },
  },
]
`;

exports[`Users [PATCH] /users/:id should update a user 1`] = `Object {}`;

exports[`Users [POST] /users should throw an error if a duplicate user is created 1`] = `
Object {
  "error": "Bad Request",
  "message": "ER_DUP_ENTRY: Duplicate entry 'test@example.com' for key 'IDX_eea9ba2f6e1bb8cb89c4e672f6'",
  "statusCode": 400,
}
`;

exports[`Users [POST] /users should throw an error if a field is the wrong format 1`] = `
Array [
  Object {
    "children": Array [],
    "constraints": Object {
      "isEmail": "emailAddress must be an email",
    },
    "property": "emailAddress",
    "target": Object {
      "emailAddress": "test[AT]example.com",
      "forename": "Duplicate",
      "surname": "User",
    },
    "value": "test[AT]example.com",
  },
]
`;

exports[`Users [POST] /users should throw an error if a required field is missing 1`] = `
Array [
  Object {
    "children": Array [],
    "constraints": Object {
      "isEmail": "emailAddress must be an email",
      "isNotEmpty": "emailAddress should not be empty",
    },
    "property": "emailAddress",
    "target": Object {
      "forename": "Duplicate",
      "surname": "User",
    },
  },
]
`;
